### ISSUE TITLE
Support for GPU (CUDA) metrics

### ISSUE THREAD

**Issue by ymotongpoo** [open-telemetry/opentelemetry-collector-contrib#392]
**Is your feature request related to a problem? Please describe.**
GPU monitoring is always left alone and requires their own hacks or dedicated tools. Having the OT integration with CUDA helps GPU users.

**Describe the solution you'd like**
Add the receiver that integrates with CUDA (NVML) and fetch metrics available from `nvidia-smi`

**Additional context**
I had researched how GPU users monitors their GPUs and found that they seldom use monitoring backend for it because of the setup cost. It's great if OT helps those users to initiate better monitoring experience.

**Comment by ymotongpoo** [comment]
Let us discuss the suggestion to have the dedicated repository for GPU metrics extractor binary as suggested in this pull request (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/393)

In this comment (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/393#issuecomment-662777421), @tigrannajaryan suggested to create an independent binary to extract GPU metrics that emits Prometheus format data so that the Collector can consume the data without change (using Prometheus receiver) to resolve this issue. The suggestion was because of CGO dependency.

As I commented in my reply (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/393#issuecomment-664685070), I think OTel can set a repository for the GPU metrics binary with the following reasons:

1. NVIDIA's GPU is defacto among famous public cloud platforms (i.e. GCP, AWS, Azure, Oracle, IBM, etc.)
2. Anyway someone needs to write a GPU metrics collection binary to integrate with OpenTelemetry Collector
3. The feature completes the gap between GPU users and existing OpenTelemetry projects

Though the binary itself is not defined in OTel spec, the objective is to fill the gap between the GPU instance monitoring and OTel and the binary is the complement part for OTel, and I think it makes sense to have the dedicated binary for it.

@mtwo @james-bebbington and @anuraaga are supporters for the idea. @tigrannajaryan @jrcamp @bogdandrutu Can you suggest the process to propose the SIG, Working Group, etc. for this? I couldn't find any from the community README.
https://github.com/open-telemetry/community

**Comment by SergeyKanzhelev** [comment]
@ymotongpoo do we need a separate SIG/WG or simply a repo to host code? Can one of existing SIGs own it?

**Comment by ymotongpoo** [comment]
Thank you @SergeyKanzhelev for the reaction and I appreciate your support! Because this is a complement feature for the Collector, WG for Collector SIG makes sense to me. Otherwise, just a simple repo should work of course.

**Comment by dashpole** [comment]
Would something like NVidia's dcgm exporter work?  https://github.com/NVIDIA/gpu-monitoring-tools.  It can attach pod info as well.

**Comment by elgalu** [comment]
@open-telemetry/collector-contrib-triagers this is still relevant, currently OpenTelemetry Collector does not support GPU related metrics however Sentry's OpenTelemetry Collector[does collect GPU metrics](https://www.sentrysoftware.com/products/hardware-sentry-opentelemetry-collector.html) , since Sentry is open source, is there a way to leverage on their work?

**Comment by elgalu** [comment]
@opentelemetry-bot /label receiver/prometheus waiting-for-author -exporter/prometheus

**Comment by fatsheep9146** [comment]
> @open-telemetry/collector-contrib-triagers this is still relevant, currently OpenTelemetry Collector does not support GPU related metrics however Sentry's OpenTelemetry Collector [does collect GPU metrics](https://www.sentrysoftware.com/products/hardware-sentry-opentelemetry-collector.html) , since Sentry is open source, is there a way to leverage on their work?

sorry, I do not find the open source for sentry to support GPU metrics. Do you know where to find this? 
@elgalu

**Comment by elgalu** [comment]
Sorry that was the wrong link. Should be https://www.sentrysoftware.com/docs/hws-otel-collector/latest/metrics.html but I also cannot find the open source code. Perhaps is closed source then:(

**Comment by MovieStoreGuy** [comment]
I've briefly looked at this since I was interested myself just randomly, the issue being that I can see is that we would need the user to either run with CGO enabled (meaning that the binaries created won't be portable since we will required additional libs to be installed) or rely on calling the underlying command (like `nvidia-smi`) which some receivers already do but it does make the portability hard in this.

**Comment by lfpalacios** [comment]
I would like to monitor GPU metrics on EKS with OpenTelemetry. Can we reopen?
Is it anything related?
https://opentelemetry.io/docs/specs/semconv/system/hardware-metrics/#hwgpu---gpu-metrics

**Comment by llc1123** [comment]
I am working on monitoring GPUs over OTel. My current approach is to use a separate process to fetch the metrics via nvml and export them in prometheus format.

If we want to do in the OTel way, we have 2 options:
- Use a separate binary as an agent, export the metrics using OTLP over HTTP/gRPC to the collector.
  - Can use nvml without a hassle.
- Implement a receiver in otel-collector.
  - Due to CGO limitations, we can only fetch via nvidia-smi cli.

I personally prefer the first approach, but that would be a project outside of OTel.

**Comment by atoulme** [comment]
@MovieStoreGuy would you have another look? Is there already a solution parsing nvidia-smi we can leverage?

**Comment by LujieDuan** [comment]
We have existing implementations of NVML and DCGM otel receivers: 
- [nvmlreceiver](https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector/tree/master/components/otelopscol/receiver/nvmlreceiver)
- [dcgmreceiver](https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector/tree/master/components/otelopscol/receiver/dcgmreceiver)

Both rely on the NVIDIA's go binding so need CGO enabled to compile. Those two receivers are included in the Ops Agent for GCP GCE customers to monitor GPUs. We couldn't upstream them because of CGO (and therefore can't be cross-compiled).  If a dedicated repository for GPU receiver is feasible, happy to work on upstreaming those GPU receivers!

**Comment by pmady** [comment]
I'd like to take a crack at building this as a `receiver/gpu` component.

This issue is almost 5 years old and the need has only gotten more urgent — with LLM inference workloads everywhere, GPU observability is no longer a nice-to-have. Right now teams are running NVIDIA's DCGM Exporter as a sidecar and scraping it with the Prometheus receiver, which works but means:

- An extra container per node just to get GPU metrics into OTel
- No native OTLP semantics — everything comes through as Prometheus metric names
- No correlation with OTel traces/logs from the same workload
- DCGM Exporter config is separate from the Collector config, so you're managing two things

A native `receiver/gpu` that calls NVML directly (same library `nvidia-smi` uses) would eliminate the sidecar, emit metrics with proper OTel resource attributes (`gpu.vendor`, `gpu.model`, `gpu.index`), and fit cleanly into existing Collector pipelines.

**Metrics I'm planning to cover (based on what NVML exposes):**

- `gpu.utilization` — GPU compute utilization %
- `gpu.memory.used` / `gpu.memory.total` — VRAM usage
- `gpu.temperature` — current temp in celsius
- `gpu.power.draw` / `gpu.power.limit` — power consumption
- `gpu.clock.sm` / `gpu.clock.memory` — current clock speeds
- `gpu.encoder.utilization` / `gpu.decoder.utilization`
- `gpu.pcie.throughput.tx` / `gpu.pcie.throughput.rx`
- `gpu.ecc.errors` — for datacenter GPUs

**Approach:**
1. Build the receiver externally first (Go, using NVML bindings via go-nvml or cgo)
2. Follow the hostmetrics receiver pattern for device-level metrics
3. Propose donation once it's stable and tested

I run GPU workloads on Kubernetes at scale (AI inference with vLLM, training jobs via Volcano) and I've been doing CNCF infrastructure work recently — contributed the `hf://` and `modelscope://` protocol backends to Dragonfly. Also speaking at HPSF 2026 on scaling Kubernetes for HPC/AI workloads, where GPU observability is always the first question people ask.

Happy to start on this if there's interest. Would love to find 2 other folks interested in co-maintaining and an approver willing to sponsor once the implementation is ready.

**Comment by braydonk** [comment]
Hi @pmady,

We built GPU receivers for our [Ops Agent](https://docs.cloud.google.com/stackdriver/docs/solutions/agents/ops-agent) product (you can read our original [blog post announcement](https://cloud.google.com/blog/products/management-tools/ops-agent-tracks-nvidia-gpus-in-compute-engine-vms) to learn more).

The receivers are open source here:
https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector/tree/master/components/otelopscol/receiver/nvmlreceiver
https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector/tree/master/components/otelopscol/receiver/dcgmreceiver

The biggest challenge with upstreaming these is the requirement for CGO. These receivers leverage the Go bindings for [libnvidia-ml.so](https://github.com/NVIDIA/go-nvml) and [libdcgm.so](https://github.com/NVIDIA/go-dcgm) developed by NVIDIA. In a blog post I wrote about a complicated bug we experienced with these libraries, you can read either the [full post](https://blog.ragecage64.com/blog/dynamic-linking-madness/) or just the section where I explain "How go-nvml works" (note to self: my blog should probably support section anchors...)

The Collector has previously not allowed components that require CGO, as it ends up spreading to the whole build needing CGO. (At one point I recall the Darwin build was CGO linked because of a hostmetrics receiver requirement that ended up getting cleared up a couple years ago, but since then it's been `CGO_ENABLED=0`).

There is perhaps a case to be made for a specific GPU distribution of the Collector that included these, and that build only could need to use CGO. But I think as long as the Collector does not intend to use CGO in builds, it won't be possible to include these GPU receivers.

If there was a path to donation though, I believe my team is fully open to exploring the options.

**Comment by pmady** [comment]
Thanks @braydonk — really appreciate the detailed context on the CGO challenge and sharing the Ops Agent receivers. I read through both the nvmlreceiver and dcgmreceiver source, and also your blog post on the dynamic linking issues — super informative.

You're right that CGO is the fundamental blocker for including these in the standard contrib build. Given that, I see a few possible paths forward:

**Option 1: GPU-specific Collector distribution**

A dedicated `opentelemetry-collector-gpu` distribution (similar to how there's a `contrib` distribution separate from `core`) that builds with `CGO_ENABLED=1` and includes the GPU receivers. This keeps CGO out of the main builds entirely. The GPU distribution would only target `linux/amd64` and `linux/arm64` (the platforms where NVIDIA GPUs actually exist), so cross-compilation isn't a concern.

**Option 2: Separate repo under opentelemetry org**

Host the GPU receivers in a dedicated `opentelemetry/opentelemetry-collector-gpu` repo (similar to how some components live outside contrib). This gives full control over the build pipeline and CGO requirements without impacting the existing collector builds.

**Option 3: Runtime dlopen without CGO**

There's been some exploration of calling NVML via `purego` or runtime `dlopen` without compile-time CGO linkage. This is what projects like [go-nvml's purego branch](https://github.com/NVIDIA/go-nvml) have experimented with. If that becomes viable, the CGO blocker goes away entirely. Though this is less mature.

I think **Option 1 or 2** is the most realistic near-term path. The Google Ops Agent receivers are already production-hardened — it would be a shame to keep reinventing this when the code exists and the team wants to donate.

I've started a prototype receiver at [pmady/otel-gpu-receiver](https://github.com/pmady/otel-gpu-receiver) to explore the metric semantics and scraper pattern, but if there's a path to upstreaming Google's existing implementations, I'd much rather help make that happen than build a third version from scratch.

**What I can help with:**
- Co-maintain the GPU receiver(s) once donated
- Help set up the GPU-specific distribution build pipeline
- Test on real GPU clusters (I run A100/H100 inference workloads on Kubernetes daily)
- Work on the OTel semantic conventions for GPU metrics (there's no official `gpu.*` namespace yet)

@braydonk @LujieDuan — would your team be interested in co-proposing a GPU distribution or separate repo? Happy to draft an OTEP or proposal if that would help move things forward. I think between Google's production implementation and real-world GPU cluster testing, we have everything needed to make this happen.

cc @atoulme @MovieStoreGuy for visibility on the path forward discussion.

**Comment by braydonk** [comment]
> This is what projects like [go-nvml's purego branch](https://github.com/NVIDIA/go-nvml) have experimented with.

Your LLM appears to have hallucinated this; I don't see a branch that makes any reference to `purego` in `go-nvml`. A code search on the repo for `purego` in the repo [provides no results](https://github.com/search?q=repo%3ANVIDIA%2Fgo-nvml%20purego&type=code).

I have experimented with `purego` for loading these libraries. It is a really big lift; you need to manually write code for all the bindings, where they are currently [generated from a header in `go-nvml`](https://github.com/NVIDIA/go-nvml?tab=readme-ov-file#how-the-bindings-are-generated). I'm also not fully convinced it is a good idea to put `purego` into everyone's production setups yet, but that is mostly a personal judgement and not based on anything I can state that is empirically wrong with it.

The main thing that would change my mind would be if NVIDIA actually did do the thing your LLM hallucinated about and started to bless it as a legitimate route for loading the shared objects.

> Host the GPU receivers in a dedicated opentelemetry/opentelemetry-collector-gpu repo (similar to how some components live outside contrib). This gives full control over the build pipeline and CGO requirements without impacting the existing collector builds.

This doesn't look to me like a separate option in itself. This is essentially just a separate question of whether the components live in contrib or not, but it doesn't answer where the build is going to live. The components on their own are just Go modules, they need to be built somehow.

> would your team be interested in co-proposing a GPU distribution or separate repo?

We are happy to donate these components if there is a clear path and consensus for how to do so.

**Comment by pmady** [comment]
@braydonk — thank you for the correction. You're absolutely right, there is no purego branch in go-nvml. That was incorrect on my part and I should have verified before citing it. I appreciate you catching that.

And fair point on Option 2 — a separate repo doesn't answer the build question by itself. The real question is: where does a CGO-enabled build pipeline live, regardless of where the component source code sits.

So it sounds like the core problem to solve is:

**How do we create an officially blessed, CGO-enabled Collector build that can include GPU receivers?**

Given your experience building and shipping these receivers in production at Google, you have a much better sense of what's realistic here than I do. A couple of questions, if you don't mind:

1. **Is a GPU-specific distribution the right framing?** i.e., an `otelcol-gpu` binary built with `CGO_ENABLED=1`, published alongside the existing `otelcol` and `otelcol-contrib` distributions — or is there a better model you'd suggest?

2. **What would the path to consensus look like?** Would this need an OTEP, a discussion in the Collector SIG, or something else? Happy to help drive whatever process makes sense, but I want to make sure we're going through the right channels.

3. **On the semantic conventions side** — did your team define GPU metric names following any particular convention, or is that still an open area? I noticed the Ops Agent receivers use their own metric naming. If upstream donation happens, aligning on `gpu.*` semantics early would be valuable.

I'd really like to help make the donation path work. Your team has the production-hardened implementation and deep expertise on the NVML/DCGM integration challenges — I can contribute testing on GPU clusters, help with the distribution build setup, and do whatever legwork is needed on the process side.

What do you think is the best next step?

**Comment by braydonk** [comment]
Hi @pmady,

I have spoken privately to some releases maintainers and they aren't open to a gpu distribution of the Collector. As a result there isn't really a super clean path to donation. You are welcome to include our components using OCB in a custom build, though we don't currently provide any support or guarantees on them outside of our own usage.

**Comment by pmady** [comment]
Thanks for the update, @braydonk. Understood on the GPU distribution — I appreciate you checking with the releases maintainers.

A quick clarification: would there be appetite for accepting the GPU receiver as a component in `opentelemetry-collector-contrib` (similar to how `hostmetricsreceiver` ships there), without requiring a separate distribution? Users could then include it via OCB or if it lands in contrib.

Either way, I'll ensure the receiver is OCB-compatible for custom builds.

**Comment by mx-psi** [comment]
You can host your component on a separate repository, see [this guide](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/docs/new-components.md#hosting-your-component-outside-of-the-opentelemetry-collector-contrib-repository). It could be picked up via OCB then. This is now a requirement before considering new components (note that, given the CGO restriction, I cannot ensure that you component would be accepted)
